keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,150 @@
+allow the user to specify whether they want the ownership change to occur.
+
+## Motivation
+
+When a volume is mounted on the node, we recursively change permissions of volume
+before bind mounting the volume inside container. The reason of doing this is to ensure
+that volumes are readable/writable by provided fsGroup.
+
+But this presents following problems:
+ - An application(many popular databases) which is sensitive to permission bits changing
+   underneath may refuse to start whenever volume being used inside pod gets mounted on
+   different node.
+ - If volume has a large number of files, performing recursive `chown` and `chmod`
+   could be slow and could cause timeout while starting the pod.
+
+### Goals
+
+ - Allow volume ownership and permission change to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+ - This proposal does not attempt to fix two pods using same volume with conflicting fsgroup. `FSGroupPermissionChangePolicy` also will be only applicable to volume types which support setting fsgroup.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on the volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `pod.Spec.SecurityContext` field expanded to include a new field called `FSGroupPermissionChangePolicy` which can have following possible values:
+
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented. Algorithm that performs permission change however will be changed to perform permission change of top level directory last.
+ - `OnDemand` --> Only perform permission and ownership change if permissions of top level directory does not match with expected permissions.
+
+```go
+type PodFSGroupPermissionChangePolicy string
+
+const(
+    OnDemandChangeVolumePermission PodFSGroupPermissionChangePolicy = "OnDemand"
+    AlwaysChangeVolumePermission PodFSGroupPermissionChangePolicy = "Always"
+)
+
+type PodSecurityContext struct {
+    // FSGroupPermissionChangePolicy ← new field
+    // + optional
+    FSGroupPermissionChangePolicy *PodFSGroupPermissionChangePolicy
+}
+```
+
+### Risks and Mitigations
+
+- One of the risks is if user volume's permission was previously changed using old algorithm(which changes permission of top level directory first) and user opts in for `OnDemand` `FSGroupPermissionChangePolicy` then we can't distinguish if the volume was previously only partially recursively chown'd.
```

Comment:
Is there a mitigation? How about remediation? Do they just need to redeploy with `Always`?

Reply:
This is a corner case. One easy mitigation is obviously to redeploy with `Always` policy but we are also considering failing pod start when permission change fails. For CSI drivers that is already the case (ie if `SetVolumeOwnership` call failed, pod can't start).